### PR TITLE
Futures netcdf no unlimited

### LINKED ISSUE #2835: Change default behaviour for FUTURE option: `netcdf_no_unlimited`
This involves completing the following work (list may not be complete):

* [x] set the default value of the FUTURE flag `netcdf_no_unlimited` to `True` in `iris.Future`.
* [x] raise an error iff a user sets a value for this FUTURE flag (that is, raise an error if the flag is set by the user, but _don't_ raise the error when the default behaviour is set up by Iris).
* [x] delete the deprecated behaviour in `iris.fileformats.netcdf.save` and `iris.fileformats.netcdf.Saver.write`, and raise an exception if a user tries to set the FUTURE flag to `False`.
* [x] update tests (including removing tests for the deprecated behaviour that was removed)
* [x] add new tests for the updated behaviour of `iris.fileformats.netcdf.save` and `iris.fileformats.netcdf.Saver.write` (if necessary – this may have been done already in the previous point) and new tests for the exception raised if the FUTURE flag is set to `False`.
* [x] Add a deprecation directive to the description of `netcdf_no_unlimited` in the docstring for `iris.Future`.
* [x] Update all doc examples to use the new behaviour

### PR DESCRIPTION
**Pull request by corinnebosley**
Closes #2835 

WIP: This is only the first few steps of the change, more to come (with appropriate fixes).

### CODE REVIEW

**Comment by dkillick** [comment]
@corinnebosley on the chance it's passed you by, I've updated the approximate list of things that need to be done in this PR on #2385 :smile:

**Comment by corinnebosley** [comment]
@dkillick Thanks, this one feels a little bit confusing.  That list is going to be essential, and much appreciated.

**Comment by corinnebosley** [comment]
Notes to self:
- Lots of cdl test failures to fix (tests expecting leading dim to be unlimited)
- Looks like maybe there aren't any unlimited dims being created at all, even on request...?
- Extra blank line in test_save on line 129
- Amend tests in iris.tests.system_test.SystemInitialTest
- Amend tests in iris.tests.unit.test_Future
- Fix this error (from test_save): TypeError: 'set' object has no attribute '__getitem__'

**Comment by corinnebosley** [comment]
@dkillick Well I'm stumped.  I really don't know why the test files don't match up, and the bits inside the @@s are just confusing and don't really tell me anything.  They seem to be suggesting that I may have moved things around in the cdl files, so now they don't match up with where things sit in the files produced in testing, but I haven't so they should.

I am annoyed.

**Comment by corinnebosley** [comment]
...maybe it's telling me that the files that are produced in the test seem to have an extra line or two at the top, so that's why they don't match up to the cdl reference files.  Let me dig into this a bit...

**Comment by corinnebosley** [comment]
@dkillick I will have to hand this over to you to finish.  Sorry.  It's got all complicated as well.

The initial test results ended up only failing a whole load of cdl match-up tests (see https://github.com/SciTools/iris/blob/master/lib/iris/tests/test_pp_cf.py#L86), so after a bit of looking we decided that this was not a problem as the variables have bits in which tell Iris what dimensions to put where, so the cubes should come out the same.

Well they don't.

The very next bit of the test I've pointed to shows that the resultant cubes don't actually match up to the reference cml files, so that sucks, and it means that something is broken.

Good luck with this one!

**Comment by dkillick** [comment]
Thanks @corinnebosley! BTW I found why all the CDL files were changing – take a look at what [assertCDL](https://github.com/SciTools/iris/blob/master/lib/iris/tests/__init__.py#L363-L369) does to CDL output...

**Comment by corinnebosley** [comment]
@dkillick Well that's sneaky.  Good spot.  Have you got Travis testing a fix for that now?

**Comment by dkillick** [comment]
@corinnebosley that's the hope...

File: lib/iris/tests/unit/test_Future.py [hunk 1]
```diff
@@ -57,6 +51,14 @@ def test_invalid_netcdf_promote_attributes(self):
             with self.assertRaisesRegexp(AttributeError, exp_emsg):
                 future.netcdf_promote = state
 
+    def test_invalid_netcdf_no_unlimited_attributes(self):
+        future = Future()
+        states = [True, False]
+        exp_emsg = ''
+        for state in states:
+            with self.assertRaisesRegexp(AttributeError, exp_emsg):
```
Comment (djkirkham) [inline lib/iris/tests/unit/test_Future.py]: You can just use `self.assertRaises()` if you don't want to specify a message.

**Comment by djkirkham** [comment]
Does this need a What's New entry?

File: lib/iris/tests/unit/test_Future.py (hunk 1 above)
Comment (dkillick) [reply lib/iris/tests/unit/test_Future.py]: No, that means I forgot to add the error message...

**Comment by dkillick** [comment]
> Does this need a What's New entry?

Yes it does, thanks @djkirkham. Will this PR never die?

File: docs/iris/src/whatsnew/contributions_2.0/incompatiblechange_2017-Oct-24_netcdf-no-unlim.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+* Deprecated FUTURE flag :attr:`iris.Future.netcdf_no_unlimited`.
+
+* Removed deprecated behaviour that automatically set the length of the outer netCDF variable to 'UNLIMITED' on save.
```
Comment (djkirkham) [inline docs/iris/src/whatsnew/contributions_2.0/incompatiblechange_2017-Oct-24_netcdf-no-unlim.txt]: Perhaps change 'con save' to 'when saving in netCDF format' or something similar, to give more context

File: docs/iris/src/whatsnew/contributions_2.0/incompatiblechange_2017-Oct-24_netcdf-no-unlim.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+  This change means that no cube dimension coordinates will be automatically saved as netCDF variables with 'UNLIMITED' length.
+  You can manually specify cube dimension coordinates to save with 'UNLIMITED' length.
```
Comment (djkirkham) [inline docs/iris/src/whatsnew/contributions_2.0/incompatiblechange_2017-Oct-24_netcdf-no-unlim.txt]: I find using the second-person a bit jarring in a technical context. Can you change this to the passive voice? Something like 'Cube dimension coordinates to save with 'UNLIMITED' length can be manually specified.'

Comment (djkirkham) [reply docs/iris/src/whatsnew/contributions_2.0/incompatiblechange_2017-Oct-24_netcdf-no-unlim.txt]: Or 'It's possible to specify...'

**Comment by pelson** [comment]
Nothing but timeouts here @djkirkham.

File: lib/iris/__init__.py [hunk 4]
```diff
@@ -210,6 +215,7 @@ def __repr__(self):
                           self.netcdf_no_unlimited, self.clip_latitudes)
 
     deprecated_options = {'cell_datetime_objects': 'warning',
+                          'netcdf_no_unlimited': 'error',
```
Comment (rhattersley) [inline lib/iris/__init__.py]: The "error" behaviour doesn't seem right. 😕
If I've been a Good User and taken account of the deprecation warnings then my current, v1.9 compatible code should look something like:
```
with iris.Future(netcdf_no_unlimited=True):
    # Do clever stuff with netCDF files
```
With the behaviour proposed in this PR, when I upgrade to v2.0 my code will get an error. 👎 The whole point of deprecations is to allow seemless migration.
Instead, I should get a deprecation warning about setting `netcdf_no_unlimited=True`. This lets my code continue to run, and I can remove the now unnecessary `with` block in my own good time.

If I've been a Bad User and ignored the deprecation warnings from v1.9, my code might just be:
```
# Do clever stuff with netCDF files that relies on the old behaviour
```
In which case my code is now broken. That's my own silly fault.
Or perhaps my code looks like:
```
with iris.Future(netcdf_no_unlimited=False):
    # Do clever stuff with netCDF files that relies on the old behaviour
```
In which case I should now get an error because I'm trying to set the context to an unsupported value. Again, this is my own silly fault.

In other words, setting netcdf_no_unlimited to True should provoke a warning, but setting it to False should raise an error.

Comment (pelson) [reply lib/iris/__init__.py]: 👍 

Indeed it should. I notice that is what is documented in the what's new too...

Comment (djkirkham) [reply lib/iris/__init__.py]: I raised the same point on another PR (https://github.com/SciTools/iris/pull/2858) and was convinced that this is acceptable because 2.0 is a major release. I haven't been following all these deprecation PRs, so I'm not sure if there are others which do the same thing.

Comment (djkirkham) [reply lib/iris/__init__.py]: The same behaviour has been added in https://github.com/SciTools/iris/pull/2848

Comment (djkirkham) [reply lib/iris/__init__.py]: Discussed offline with @pelson: I'll put up a separate PR to change the behaviour of the `Future` object so that the 'error' behaviour is to raise an exception if the flag is set to `False` and raise a warning when it is set to `True`.
